kernel/hostfile.c

```diff
@@ -115,3 +139,123 @@ int myst_load_host_file(const char* path, void** data_out, size_t* size_out)
 
     return ret;
 }
+
+int myst_copy_host_directory_recursively(
+    const char* src_dir,
+    const char* dst_dir,
+    bool ignore_errors)
+{
+    int ret = 0;
+    int fd = -1;
+    int new_fd = -1;
+    long nread;
+    char d_type;
+    void* buf = NULL;
+    size_t buf_size;
+    struct linux_dirent64* d;
+    struct locals
+    {
+        char buf[BUFSIZ];
+        char symlink_target[BUFSIZ];
+        char src_path[PATH_MAX];
+        char dst_path[PATH_MAX];
+    };
+    struct locals* locals = NULL;
+
+    if (!src_dir || !dst_dir)
+        ERAISE(-EINVAL);
+
+    ECHECK(fd = _host_open(src_dir, O_RDONLY | O_DIRECTORY, 0));
+
+    if (!(locals = calloc(1, sizeof(struct locals))))
+        ERAISE(-ENOMEM);
+
+    ECHECK(myst_mkdirhier(dst_dir, 07555));
+
+    for (;;)
+    {
+        ECHECK(nread = _host_getdents64(fd, locals->buf, BUFSIZ));
+
+        if (nread == 0)
+            break;
+
+        for (long bpos = 0; bpos < nread;)
+        {
+            d = (struct linux_dirent64*)(locals->buf + bpos);
+            bpos += d->d_reclen;
+
+            if (strcmp(d->d_name, ".") == 0 || strcmp(d->d_name, "..") == 0)
+                continue;
+
+            d_type = d->d_type;
+            if (d_type != DT_DIR && d_type != DT_REG && d_type != DT_LNK)
+                continue;
+
+            ECHECK(myst_make_path(
+                locals->src_path,
+                sizeof(locals->src_path),
+                src_dir,
+                d->d_name));
+            ECHECK(myst_make_path(
+                locals->dst_path,
+                sizeof(locals->dst_path),
+                dst_dir,
+                d->d_name));
+            if (d_type == DT_DIR)
+            {
+                ECHECK(myst_copy_host_directory_recursively(
+                    locals->src_path, locals->dst_path, ignore_errors));
+            }
+            else if (d_type == DT_REG)
+            {
+                if (myst_load_host_file(locals->src_path, &buf, &buf_size) < 0)
+                {
+                    if (ignore_errors)
```

Comment:
For my understanding, `ignore_errors` only tolerates errors during copying of regular files? Is this a scenario you saw happen locally?

Reply:
That's right. For example reading `/sys/devices/system/cpu/cpu0/power/autosuspend_delay_ms` will result in an Input/output error. According to https://www.kernel.org/doc/Documentation/ABI/testing/sysfs-devices-power

> Not all drivers support this attribute.  If it isn't supported,
> attempts to read or write it will yield I/O errors.